### DISCUSSION TITLE
Help: Upload instructions for custom rnode firmware

### DISCUSSION THREAD

**Discussion by ghost** [markqvist/Reticulum#477] [Help & Questions]
# Main goal
Make the rnode firmware compatible with diy hardware that contains the ebyte e22-900m30s 1w LoRa moden. In my case it's the hydra board: https://github.com/Hydra-Designs/project-hydra-meshtastic-pcb
My main idea was to bootstrap from the existing "generic esp32" device since it's technically the same thing, but different pins and different RF chipset/rxtx switching compared to the original design. 
## Current state of my efforts
my repo: https://github.com/azidoazid/RNode_Firmware
* Manual RX/TX switching implemented, borrowed from the sx128x driver
* Custom board definition for the hydra board
* Makefile: firmware and upload section implemented, borrowed from the "generic esp32" target, modified compile flags for different board ID and radio
*  I can compile custom firmware
* upload does not work
## Error messages
* make upload fails with: `This device has not been provisioned yet, cannot set firmware hash`
* OLED shows: `Hardware Failure`
## Steps to reproduce
* `make firmware-mt_hydra`
* `make upload-mt_hydra`

# Question
* What steps are required to go from sourcecode to working device?

I have only limited coding experience, but as far as I understand there is some post-upload process involved that generates some checksums that are stored in flash memory.  I also got the same error message on the OLED while using rnodeconf to install firmware on supported hardware, but only for a short period so there must be some step that I am missing.

**Comment by cobraPA** [comment]
Hopefully Mark will drop in and give pointers to 'a better way'...
There are firmware CRC checks and signing checks done for standard installs.  For the BLE test version I am working on for iPhone, you can have a look at the (hacky) overrides I am using.
https://github.com/cobraPA/RNode_Firmware_iOS/blob/kb_ble_test1/RNode_Firmware.ino
1. Line 1153 with the "if(1)" replacement of the initial check.
2. Line 1169 and 1172 for startup check overrides
3. Might need 1225 also

Depends on what is installed and working on your board.  The printouts in that section might help you when you try to enable your Lora radio.  One of those sections will fail if the radio init fails.

The above should let you run your 'make upload-xxxx' command.  It will normally throw an error for the second step, but the two flashing phases before and after should run.  The overrides in the firmware I showed will then allow the firmware to boot without failing the CRC and signing checks which you may not have in eeprom yet.

To fully provision a radio locally, you will need to sign the firmware with rnodeconf.  But I think you also need to modify rnodeconf to add proper product IDs, if you have an entirely new board type.

**Reply by ghost** [reply]
Thanks for the quick reply! I initially forgot to set the proper SPI pins and got the error message "radio not found" on the OLED, so I am quite sure that at least my sx126x driver hacking and board definition seems to be working, at least the communication part. 
I thought that I could work off the existing "generic esp32" and just add some switches that trick the whole ecosystem into thinking that it's a "generic esp32" target (which is seemingly implemented in rnodeconf) while under the hood it's running the sx126x driver with my rxtx switching hacks.
I even tried to recycle the generic esp32 board ID but no luck :(
I am going to give your overrides a try!

**Reply by cobraPA** [reply]
There is also a recent change by Mark to help with some of this, I think:
https://github.com/markqvist/RNode_Firmware/commit/854b345e3d97f22d0bb0f1f464de9625d5ed7dea
It is not in my tree yet, because I haven't merged it, but it could also help.  I think my overrides will already override that piece though.

**Reply by markqvist** [reply]
See my full reply below, but just wanted to note here too, that I think you should definitely define your own correct `product` and `model` codes from the beginning. Just choose something that doesn't collide with anything else. You will need to add your product definition to the [eeprom_product_valid](https://github.com/azidoazid/RNode_Firmware/blob/master/Utilities.h#L1288) function and likewise add required logic to the [eeprom_model_valid](https://github.com/azidoazid/RNode_Firmware/blob/master/Utilities.h#L1300) function. Other than that, I think you already have the idea of it.

**Reply by ghost** [reply]
Gave it a try, no luck :(

**Reply by cobraPA** [reply]
What happened?  Oh, the provisioning errors?
Is there documentation on the provisioning commands?
I did go through rnodeconf... you need something like these commands the first time, unless you can flash your board with a known config first:

rnodeconf --key
rnodeconf --platform ESP32 --product f0 --model ff --hwrev 1 --rom  serial_port
  product and model would come from those functions Mark showed
  hwrev I assume is based on your product
  where serial_port is a platform style port /dev/usbxx or COM3, etc

**Comment by markqvist** [comment]
The firmware won't boot without a valid EEPROM config, but since `rnodeconf` doesn't have any idea about your board model and product codes yet, there's no easy way to actually write the config, which is basically the gist of the problem.

The `rnodeconf` utility has the `--rom` option, which allows manual EEPROM bootstrap, but I just checked, and to my actual own surprise, I found that you can't just pass any codes to it, but it was limited to a few. Hm, I did write that code a long time ago, so there's to that.

Either way, I just [pushed a new commit](https://github.com/markqvist/Reticulum/commit/e5427d70aca20633a9682a33f796488d796ca8cc) to `rnodeconf`, that will allow arbitrary EEPROM bootstrap values. You'll have to set these correctly once the firmware has been built and flashed via the makefile (when the device says "Hardware failure"). You will have to do something like:

```sh
rnodeconf --rom --product f0 --model 43 --hwrev 1
```

The actual product and model codes will have to match what you used in the firmware, though! I just guessed from a quick look at your repo, but you may or may not have changed them or set them to something else. For the final version, choose a `product` and `model` code that doesn't collide with any other boards already defined (so we can add a nice description and such to `rnodeconf` and other tools for displaying your board info).

I hope that gets your firmware running correctly :)

**Reply by ghost** [reply]
running your new rnodeconf.py version:
```
rnodeconf --rom --product f0 --model 43 --hwrev 1 /dev/ttyUSB0
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
Waiting for ESP32 reset...
An error occurred while writing EEPROM. The contained exception was:
'bytes' object cannot be interpreted as an integer
Traceback (most recent call last):
  File "/home/azidoazid/reticulum/bin/rnodeconf", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/azidoazid/reticulum/lib/python3.11/site-packages/RNS/Utilities/rnodeconf.py", line 3247, in main
    raise e
  File "/home/azidoazid/reticulum/lib/python3.11/site-packages/RNS/Utilities/rnodeconf.py", line 3094, in main
    info_chunk  = b"" + bytes([mapped_product, model, ord(hwrev)])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'bytes' object cannot be interpreted as an integer`
```
So far I've just included a new `#define BOARD_MT_HYDRA      0x43`,  nothing else.
Glad I can help to sqash some bugs! The out of the box experience with supported hardware was really awesome, I was up and running in a couple of minutes :+1:

**Reply by cobraPA** [reply]
Great, thanks Mark.  I meant to post about that thing with the 'no arbitrary values' to bootstrap firmware.  I ran in to it also.

**Reply by markqvist** [reply]
Ah, was a bit too fast there and forgot to wrap the statements in `ord()` of course. It should work now without any `TypeErrors` :)

@cobraPA: Yeah, I actually thought I'd set up like that from the beginning, but apparently not :P Good that it's there now, at least. Should make it a bit easier for everyone.

**Reply by ghost** [reply]
now we are getting closer:
```
rnodeconf --rom --product f0 --model 43 --hwrev 1 /dev/ttyUSB0
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
Waiting for ESP32 reset...
Loading signing key...
Bootstrapping device EEPROM...
EEPROM written! Validating...
Waiting for ESP32 reset...
Invalid EEPROM data, could not parse device EEPROM.
EEPROM was written, but validation failed. Check your settings.
```

After setting `#define VALIDATE_FIRMWARE false` in `device.h`:
```
rnodeconf --rom --product f0 --model 43 --hwrev 1 /dev/ttyUSB0
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
Waiting for ESP32 reset...
Loading signing key...
Bootstrapping device EEPROM...
EEPROM written! Validating...
Waiting for ESP32 reset...
Invalid EEPROM data, could not parse device EEPROM.
The contained exception was: 67
EEPROM was written, but validation failed. Check your settings.
<RNS.Utilities.rnodeconf.RNode object at 0x7f7b88aaed10> hardware error (code 03)
[bunch of the same hardware error messages]
<RNS.Utilities.rnodeconf.RNode object at 0x7f7b88aaed10> hardware error (code 03)
EEPROM written! Validating...
<RNS.Utilities.rnodeconf.RNode object at 0x7f7b88aaed10> hardware error (code 03)
[bunch of the same hardware error messages]
<RNS.Utilities.rnodeconf.RNode object at 0x7f7b88aaed10> hardware error (code 03)
Waiting for ESP32 reset...
Invalid EEPROM data, could not parse device EEPROM.
The contained exception was: 67
EEPROM was written, but validation failed. Check your settings.
```

=> OLED still shows "hardware failure"

**Reply by markqvist** [reply]
At this point, setting `VALIDATE_FIRMWARE` will have no bearing on the result - it's dealing with the firmware hash.

As I said earlier, you need to actually add your correct product and model codes to the firmware (in the locations I mentioned), and make sure those are the codes you are writing with the bootstrap command. Otherwise validation will fail *in the firmware*, and it will spit out the hardware error codes on the OLED and over serial.

For more informative info on the EEPROM error though, you could also add `RNS.trace_exception(e)` after line 950 in `rnodeconf.py`.

**Reply by markqvist** [reply]
Ok, I realised that we also probably don't need to crap out `rnodeconf` if the EEPROM data contains an unknown model code here, so I just pushed [this commit](https://github.com/markqvist/Reticulum/commit/b977f33df6d34cb88fa72c9d6c2a9e24a8399c0a) which should add a more informative error code instead, and allow EEPROM validation to continue, even though it contains an unknown model code.

Your model codes and band capabilities need to be added [to this array](https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L221) in the end, but for now, the above commit should let you get everything running and working for testing and such.

**Reply by ghost** [reply]
I am sorry but I don't understand what model and product code I have to set. It's quite confusing for someone that has no background in embedded software.
In my code I've defined a new model:
https://github.com/azidoazid/RNode_Firmware/blob/dbfb722b394cf0aca6f838d8a04b503c85f5f123/Boards.h#L42
So that I can define a custom pinout, modem and modem config (manual rx/tx switching).
I then borrowed the compile instructions from the generic esp32 variant: https://github.com/azidoazid/RNode_Firmware/blob/dbfb722b394cf0aca6f838d8a04b503c85f5f123/Makefile#L109
and modified the compile flags in order to compile in my custom pinout, modem, ... stuff (dboard and dmodem)

=> this compiles working code but rnodeconf obviously doesn't know anything about my 0x43 board ID: ` #define BOARD_MT_HYDRA      0x43`

So now my idea was to borrow the hmbrw ID 0x32
=> https://github.com/markqvist/RNode_Firmware/commit/b7fab351f99272f8cb2bfcbc2ed7d41f304da43e
But what `--product` and `--model ID` do I have to set for `rnodeconf`?
Is `--product` the  hmbrw ID 0x32? and  `--model ID`  0xA9 since I am using rnode_firmware.hex, the sx1262 and 868MHz band?

After running make firmware && make upload on my commit linked above I get:
`This device has not been provisioned yet, cannot set firmware hash`
as usual.
I then run your latest rnodeconf with:
`rnodeconf --rom --product 32 --model a9 --hwrev 1 /dev/ttyUSB0`
which results in:
```
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
Waiting for ESP32 reset...
Loading signing key...
Bootstrapping device EEPROM...
EEPROM written! Validating...
Waiting for ESP32 reset...
EEPROM checksum correct
Device signature validated
EEPROM Bootstrapping successful!

```

But the OLED still shows "Hardware Failure" :( 
Am I missing something? Compared to the regular --autoinstall method I have the feeling that there is a second upload step afterwards?

When I re-run the previously failing `rnodeconf /dev/ttyUSB0 --firmware-hash $(./partition_hashes ./build/esp32.esp32.esp32/RNode_Firmware.ino.bin)`
I get the following output:
```
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
EEPROM checksum correct
Device signature validated
Firmware hash set
```
But I still get "hardware failure" on the OLED.

Running `rnodeconf --info`:
```
Opening serial port /dev/ttyUSB0...
Device connected
Current firmware version: 1.70
Reading EEPROM...
EEPROM checksum correct
Device signature validated

Device info:
Traceback (most recent call last):
  File "/home/ozon/Dokumente/reticulum/bin/rnodeconf", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/ozon/Dokumente/reticulum/lib/python3.11/site-packages/RNS/Utilities/rnodeconf.py", line 2958, in main
    RNS.log("\tProduct            : "+products[rnode.product]+" "+models[rnode.model][3]+" ("+bytes([rnode.product]).hex()+":"+bytes([rnode.model]).hex()+board_string+")")
                                      ~~~~~~~~^^^^^^^^^^^^^^^
KeyError: 50
```


I think the best approach would be a better homebrew variant that adds support for the sx1262 and manual rx/tx switching instead of adding a totally new board. Since I don't have an esp32 devboard and SX1276 module I can't follow the diy rnode guide and build my way up from the existing implementation.

**Reply by markqvist** [reply]
Yeah, I might have misunderstood you here, since I thought you wanted to add general support for the Hydra board to the firmware and related utilities, which is why I tried pointing you in the direction of getting the model codes and such set up. If this is just for one-off use and personal experimentation, it is probably much easier for you to just modify the generic ESP32 profile and be done with it. You can just flash the board once via the autoinstaller and the generic ESP32, which will set up the EEPROM and everything, and then flash your own custom firmware via the makefile.

**Reply by cobraPA** [reply]
@azidoazid , one thing I forgot to mention, the display ‘hardware failure’ is not fatal.  It includes checksum and product errors, but does not prevent the RNode from running normally (except for preventing full display of stats.)  I think you might be focused on that error…. I had a fully running RNode with BLE and Lora, even when that was displayed.

**Reply by cobraPA** [reply]
Although, I may have bypassed one of the startup checks to get past that also.

**Reply by liamcottle** [reply]
@cobraPA:

> I had a fully running RNode with BLE and Lora, even when that was displayed.

Could you please give me any pointers on how you were using BLE?

I see a [new commit](https://github.com/markqvist/RNode_Firmware/commit/31540410cfaac2848240c0e95e3408d9a7f1bc4c) about BLE support and would love to know how to try this out.

I compiled the latest RNode_Firmware commit, and flashed it to my node, even tried using the `--bluetooth-on` and `--bluetooth-pair` flags via latest commit of `rnodeconf.py`, but I never see any bluetooth devices show up in device settings on my iOS device or Android device that look to be the node.

I'm doing this with a Heltec v3 board btw. I understand this board is only very recently in the process of being supported, but just wondering if this already works and how to go about using it with BLE.

For extra info, this is how I flashed latest commits;

```
# clone latest RNode Firmware
git clone https://github.com/markqvist/RNode_Firmware
cd RNode_Firmware

# build for heltec v3
make clean
make prep-esp32 
make firmware-heltec32_v3

# flash it
arduino-cli upload -p /dev/cu.usbserial-0001 --fqbn esp32:esp32:heltec_wifi_lora_32_V3
python3 ~/Desktop/Reticulum/RNS/Utilities/rnodeconf.py /dev/cu.usbserial-0001 --firmware-hash $(./partition_hashes ./build/esp32.esp32.heltec_wifi_lora_32_V3/RNode_Firmware.ino.bin)
python ./Release/esptool/esptool.py --chip esp32-s3 --port /dev/cu.usbserial-0001 --baud 921600 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 80m --flash_size 4MB 0x210000 ./Release/console_image.bin

# enable bluetooth????
python3 ~/Desktop/Reticulum/RNS/Utilities/rnodeconf.py /dev/cu.usbserial-0001 --bluetooth-on
python3 ~/Desktop/Reticulum/RNS/Utilities/rnodeconf.py /dev/cu.usbserial-0001 --bluetooth-pair
```

**Reply by cobraPA** [reply]
Hi @liamcottle,

Well, not short answer... but here goes.

In device settings on your iOS device, you will not see BLE devices before the are connected or paired (at least custom ones)!  I learned this early on and it was a lot of confusion initially.  This is a difference from classic Bluetooth where you commonly pair from the device list. I would suspect Android is the same but I may be wrong there.  BLE is kind of a different beast when it comes to bluetooth.  It is often used for devices that are custom to the application they attach to, so the applications are expected to know how to identify them, either with their own custom type or something standard.  If you want to be able to scan for them, on iOS there is a good BLE diagnostic app called LightBlue.

For Android with Heltec V3 you will need to use a serial connection for now.  For iOS, if you install the test SidebandOS app and then flash one of your RNodes with https://github.com/cobraPA/RNode_Firmware_iOS/tree/kb_ble_test1, then you can connect it.

**Reply by liamcottle** [reply]
@cobraPA Perfect! Thanks, that was the info I needed. I was mistaken thinking those recent BLE commits to master firmware included this. After flashing the firmware from your `kb_ble_test1` branch, against all odds, this worked for me.

I was able to disable WiFi and Cellular on my iOS device, launch Sideband, pair with the Heltec v3 RNode via BLE and then send a message from Sideband to my Windows PC that has another RNode connected to it via USB (running Sideband as well).

Great work on this so far! There are a few bugs I noticed, but it works! I noticed if you minimise the iOS app at all, Sideband disconnects from the RNode. You can see on the RNode screen that the BLE and LoRa icons change to disconnected state, and the RF traffic graph stops moving. I am unable to get the app to auto reconnect at all, by toggling any bluetooth settings etc... However, if I force close Sideband and relaunch it, it does in fact reconnect to BLE and everything starts working again!

One other thing I wanted to mention, is it possible to update the Sideband version on Test Flight? Currently it's on v0.7.7 and is missing the functionality to send attachments/images. These features exist in the v0.8.3 version I'm using on Windows/Android. It might be painful to do so since you have these in-development changes for BLE, so no problem if you can't do so easily.

Happy to test changes and report back if needed.

Thanks again!

**Reply by cobraPA** [reply]
Thanks for checking it out.  Yes, updating to newer version of Sideband is on my todo.  It's usually a large merge, although not always a hard one.  The image send may take more work to hook in to iOS though.  I also want to get the telemetry/gps integration working.  You should find that the map displays, but the gps is not wired in on iOS yet.
There is some in-progress work on BLE background operation for the RNode.  It should function for a short time at least, but more testing to be done there.  I believe with the proper setup that it should work in the background, as supporting custom BLE devices is one of the allowed background modes for iOS (but that is not all done yet.)
